File: intrinsic/motion_planning/motion_planner_client.cc
// Copyright 2023 Intrinsic Innovation LLC

#include "intrinsic/motion_planning/motion_planner_client.h"

#include <memory>
#include <optional>
#include <string>
#include <utility>
#include <vector>

#include "absl/status/statusor.h"
#include "google/protobuf/duration.pb.h"
#include "grpcpp/client_context.h"
#include "intrinsic/eigenmath/types.h"
#include "intrinsic/icon/proto/cart_space_conversion.h"
#include "intrinsic/icon/proto/joint_space.pb.h"
#include "intrinsic/icon/release/status_helpers.h"
#include "intrinsic/logging/proto/context.pb.h"
#include "intrinsic/math/proto_conversion.h"
#include "intrinsic/motion_planning/conversions.h"
#include "intrinsic/motion_planning/proto/motion_planner_config.pb.h"
#include "intrinsic/motion_planning/proto/motion_planner_service.grpc.pb.h"
#include "intrinsic/motion_planning/proto/motion_planner_service.pb.h"
#include "intrinsic/motion_planning/proto/motion_target.pb.h"
#include "intrinsic/util/eigen.h"
#include "intrinsic/world/objects/kinematic_object.h"

namespace intrinsic {
namespace motion_planning {

const MotionPlannerClient::MotionPlanningOptions&
MotionPlannerClient::MotionPlanningOptions::Defaults() {
  static const auto* defaults = new MotionPlannerClient::MotionPlanningOptions({
      .path_planning_time_out = 30,
      .compute_swept_volume = false,
  });

  return *defaults;
}

MotionPlannerClient::MotionPlannerClient(
    absl::string_view world_id,
    std::shared_ptr<
        intrinsic_proto::motion_planning::MotionPlannerService::StubInterface>
        motion_planner_service)
    : world_id_(world_id),
      motion_planner_service_(std::move(motion_planner_service)) {}

absl::StatusOr<MotionPlannerClient::PlanTrajectoryResult>
MotionPlannerClient::PlanTrajectory(
    const intrinsic_proto::motion_planning::RobotSpecification&
        robot_specification,
    const intrinsic_proto::motion_planning::MotionSpecification&
        motion_specification,
    const MotionPlanningOptions& options, const std::string& caller_id,
    const intrinsic_proto::data_logger::Context& context) {
  intrinsic_proto::motion_planning::MotionPlanningRequest request;
  *request.mutable_robot_specification() = robot_specification;
  *request.mutable_motion_specification() = motion_specification;
  request.set_world_id(world_id_);
  request.set_compute_swept_volume(options.compute_swept_volume);
  request.mutable_motion_planner_config()->mutable_timeout_sec()->set_seconds(
      options.path_planning_time_out);
  const int64_t s = request.motion_planner_config().timeout_sec().seconds();
  request.mutable_motion_planner_config()->mutable_timeout_sec()->set_nanos(
      (options.path_planning_time_out - s) * 1e9);
  request.set_caller_id(caller_id);
  *request.mutable_context() = context;

  intrinsic_proto::motion_planning::TrajectoryPlanningResponse response;
  grpc::ClientContext ctx;
  INTRINSIC_RETURN_IF_ERROR(
      motion_planner_service_->PlanTrajectory(&ctx, request, &response));

  MotionPlannerClient::PlanTrajectoryResult result;
  result.trajectory = response.discretized();
  result.swept_volume.insert(result.swept_volume.begin(),
                             response.swept_volume().begin(),
                             response.swept_volume().end());

  return result;
}

absl::StatusOr<std::vector<eigenmath::VectorXd>> MotionPlannerClient::ComputeIk(
    const world::KinematicObject& robot,
    const intrinsic_proto::motion_planning::CartesianMotionTarget&
        cartesian_target,
    const IkOptions& options) {
  intrinsic_proto::motion_planning::IkRequest request;
  request.set_world_id(world_id_);
  request.mutable_robot_reference()->mutable_object_id()->set_id(
      robot.Id().value());
  *request.mutable_target() = cartesian_target;

  if (options.starting_joints.size() > 0) {
    VectorXdToRepeatedDouble(
        options.starting_joints,
        request.mutable_starting_joints()->mutable_joints());
  }

  if (options.max_num_solutions.has_value()) {
    request.set_max_num_solutions(*options.max_num_solutions);
  }

  if (options.collision_settings.has_value()) {
    *request.mutable_collision_settings() = *options.collision_settings;
  }

  request.set_ensure_same_branch(options.ensure_same_branch);

  intrinsic_proto::motion_planning::IkResponse response;
  grpc::ClientContext ctx;
  INTRINSIC_RETURN_IF_ERROR(
      motion_planner_service_->ComputeIk(&ctx, request, &response));

  return ToVectorXds(response.solutions());
}

namespace {

// Common adaptor to handle different specifications of reference, target.
absl::StatusOr<Pose3d> ComputeFkInternal(
    const world::KinematicObject& robot,
    const eigenmath::VectorXd& joint_values,
    const intrinsic_proto::world::TransformNodeReference& reference,
    const intrinsic_proto::world::TransformNodeReference& target,
    const std::string& world_id,
    intrinsic_proto::motion_planning::MotionPlannerService::StubInterface&
        motion_planner_service) {
  intrinsic_proto::motion_planning::FkRequest request;
  request.set_world_id(world_id);
  request.mutable_robot_reference()->mutable_object_id()->set_id(
      robot.Id().value());
  VectorXdToRepeatedDouble(joint_values,
                           request.mutable_joints()->mutable_joints());

  *request.mutable_reference() = reference;
  *request.mutable_target() = target;

  intrinsic_proto::motion_planning::FkResponse response;
  grpc::ClientContext ctx;
  INTRINSIC_RETURN_IF_ERROR(
      motion_planner_service.ComputeFk(&ctx, request, &response));

  return FromProto(response.reference_t_target());
}

}  // namespace

absl::StatusOr<Pose3d> MotionPlannerClient::ComputeFk(
    const world::KinematicObject& robot,
    const eigenmath::VectorXd& joint_values,
    const intrinsic_proto::world::TransformNodeReferenceByName& reference,
    const intrinsic_proto::world::TransformNodeReferenceByName& target) {
  intrinsic_proto::world::TransformNodeReference reference_proto;
  *reference_proto.mutable_by_name() = reference;
  intrinsic_proto::world::TransformNodeReference target_proto;
  *target_proto.mutable_by_name() = target;
  return ComputeFkInternal(robot, joint_values, reference_proto, target_proto,
                           world_id_, *motion_planner_service_);
}

absl::StatusOr<Pose3d> MotionPlannerClient::ComputeFk(
    const world::KinematicObject& robot,
    const eigenmath::VectorXd& joint_values,
    const world::TransformNode& reference, const world::TransformNode& target) {
  intrinsic_proto::world::TransformNodeReference reference_proto;
  reference_proto.set_id(reference.Id().value());
  intrinsic_proto::world::TransformNodeReference target_proto;
  target_proto.set_id(target.Id().value());
  return ComputeFkInternal(robot, joint_values, reference_proto, target_proto,
                           world_id_, *motion_planner_service_);
}

absl::StatusOr<intrinsic_proto::motion_planning::CheckCollisionsResponse>
MotionPlannerClient::CheckCollisions(
    const world::KinematicObject& robot,
    const std::vector<eigenmath::VectorXd>& waypoints,
    const CheckCollisionsOptions& options) {
  intrinsic_proto::motion_planning::CheckCollisionsRequest request;
  request.set_world_id(world_id_);
  request.mutable_robot_reference()->mutable_object_id()->set_id(
      robot.Id().value());
  ToJointVecs(waypoints, request.mutable_waypoint());

  if (options.collision_settings.has_value()) {
    *request.mutable_collision_settings() = *options.collision_settings;
  }

  intrinsic_proto::motion_planning::CheckCollisionsResponse response;
  grpc::ClientContext ctx;
  INTRINSIC_RETURN_IF_ERROR(
      motion_planner_service_->CheckCollisions(&ctx, request, &response));
  return response;
}

absl::StatusOr<intrinsic_proto::motion_planning::ClearCacheResponse>
MotionPlannerClient::ClearCache() {
  intrinsic_proto::motion_planning::ClearCacheResponse response;
  grpc::ClientContext ctx;
  INTRINSIC_RETURN_IF_ERROR(motion_planner_service_->ClearCache(
      &ctx, intrinsic_proto::motion_planning::ClearCacheRequest(), &response));
  return response;
}

}  // namespace motion_planning
}  // namespace intrinsic

File: intrinsic/motion_planning/motion_planner_client.py
# Copyright 2023 Intrinsic Innovation LLC

"""Defines the MotionPlannerClient class.

The MotionPlannerClient provides access to computations on top of the world.
This includes:
  * IK/FK
  * path planning
"""

import dataclasses
from typing import List, Optional

# This import is required to use the *_grpc imports.
# pylint: disable=unused-import
import grpc

from google.protobuf import empty_pb2
from intrinsic.icon.proto import joint_space_pb2
from intrinsic.math.python import data_types
from intrinsic.math.python import proto_conversion as math_proto_conversion
from intrinsic.motion_planning.proto import motion_planner_config_pb2
from intrinsic.motion_planning.proto import motion_planner_service_pb2
from intrinsic.motion_planning.proto import motion_planner_service_pb2_grpc
from intrinsic.motion_planning.proto import motion_specification_pb2
from intrinsic.motion_planning.proto import motion_target_pb2
from intrinsic.motion_planning.proto import robot_specification_pb2
from intrinsic.world.proto import collision_settings_pb2
from intrinsic.world.proto import geometric_constraints_pb2
from intrinsic.world.python import object_world_ids


def _repeated_vec_to_list_of_floats(
    vectors: List[joint_space_pb2.JointVec],
) -> List[List[float]]:
  return [list(vector.joints) for vector in vectors]


@dataclasses.dataclass
class CheckCollisionsOptions:
  """Options for Collision settings.

  Attributes:
    collision_settings: Settings for collision checking.
  """

  collision_settings: collision_settings_pb2.CollisionSettings | None = None


@dataclasses.dataclass
class IKOptions:
  """Options for IK.

  Attributes:
    max_num_solutions: The maximum number of solutions to be returned from IK
      computation. If not set (== 0), the underlying implementation has the
      freedom to choose. Negative values are invalid.
    starting_joints: The starting joint configuration to use. If not set, the
      current position of the robot in the world will be used.
    collision_settings: Collision Settings. If left empty, no collision checking
      is done.
    ensure_same_branch: Flag to choose IK solution is on the same kinematic
      branch as the starting joints of the robot.
    prefer_same_branch:  Flag that will prefer solutions on the same kinematic
      branch over those close to the starting_joints configuration.
  """

  max_num_solutions: int = 0
  starting_joints: list[float] | None = None
  collision_settings: collision_settings_pb2.CollisionSettings | None = None
  ensure_same_branch: bool = False
  prefer_same_branch: bool = False


@dataclasses.dataclass
class MotionPlanningOptions:
  """Options for Motion Planning.

  Attributes:
    path_planning_time_out: Timeout for path planning algorithms.
    lock_motion_configuration: Optional configuration for saving or loading a
      motion.
  """

  path_planning_time_out: int = 30
  lock_motion_configuration: (
      motion_planner_config_pb2.LockMotionConfiguration | None
  ) = None


class MotionPlannerClient:
  """Helper class for calling the rpcs in the MotionPlannerService.

  Provides additional computations on top of the world.
    * IK/FK
    * Path planning
  """

  def __init__(
      self,
      world_id: str,
      stub: motion_planner_service_pb2_grpc.MotionPlannerServiceStub,
  ):
    self._world_id: str = world_id
    self._stub: motion_planner_service_pb2_grpc.MotionPlannerServiceStub = stub

  def plan_trajectory(
      self,
      robot_specification: robot_specification_pb2.RobotSpecification,
      motion_specification: motion_specification_pb2.MotionSpecification,
      options: MotionPlanningOptions = MotionPlanningOptions(),
      caller_id: str = "Anonymous",
  ) -> joint_space_pb2.JointTrajectoryPVA:
    """Plan trajectory for a given motion planning problem and robot.

    This method calls the Plan trajectory rpc.

    Args:
      robot_specification: Robot specification
      motion_specification: Motion specification, see MotionSpecification proto.
      options: Motion planning options that allows the path planning timeout to
        be set.
      caller_id: The id used for logging the request in the motion planner
        service.

    Returns:
      Discretized trajectory
    """
    request = motion_planner_service_pb2.MotionPlanningRequest(
        world_id=self._world_id,
        robot_specification=robot_specification,
        motion_specification=motion_specification,
        caller_id=caller_id,
    )
    request.motion_planner_config.timeout_sec.seconds = (
        options.path_planning_time_out
    )
    if options.lock_motion_configuration:
      request.motion_planner_config.lock_motion_configuration.CopyFrom(
          options.lock_motion_configuration
      )
    response = self._stub.PlanTrajectory(request)
    return response.discretized

  def compute_ik(
      self,
      robot_name: object_world_ids.WorldObjectName,
      target: (
          motion_target_pb2.CartesianMotionTarget
          | geometric_constraints_pb2.GeometricConstraint
      ),
      starting_joints: Optional[List[float]] = None,
      options: Optional[IKOptions] = None,
  ) -> List[List[float]]:
    """Calls the ComputeIk rpc, doing argument conversion as necessary.

    Args:
      robot_name: Name of robot, must map to a kinematic object
      target: a target pose to compute ik for
      starting_joints: used as seed for ik, optional
      options: See IKOptions.

    Returns:
      A list of IK solutions
    """
    request = motion_planner_service_pb2.IkRequest(world_id=self._world_id)
    request.robot_reference.object_id.by_name.object_name = robot_name

    # Convert CartesianMotionTarget to GeometricConstraint::PoseEquality.
    if isinstance(target, motion_target_pb2.CartesianMotionTarget):
      request.target.pose_equality.moving_frame.CopyFrom(target.tool)
      request.target.pose_equality.target_frame.CopyFrom(target.frame)
      if target.HasField("offset"):
        request.target.pose_equality.target_frame_offset.CopyFrom(target.offset)
    else:
      request.target.CopyFrom(target)

    if starting_joints is not None:
      request.starting_joints.joints.extend(starting_joints)
    if options is not None:
      if options.collision_settings:
        request.collision_settings.CopyFrom(options.collision_settings)
      if options.ensure_same_branch:
        request.ensure_same_branch = options.ensure_same_branch
      if options.prefer_same_branch:
        request.prefer_same_branch = options.prefer_same_branch
      if options.max_num_solutions:
        request.max_num_solutions = options.max_num_solutions

    # Make the rpc.
    response = self._stub.ComputeIk(request)
    return _repeated_vec_to_list_of_floats(response.solutions)

  def compute_fk(
      self,
      robot_name: object_world_ids.WorldObjectName,
      joints: List[float],
      reference: object_world_ids.WorldObjectName,
      target: object_world_ids.WorldObjectName,
      reference_frame: Optional[object_world_ids.FrameName] = None,
      target_frame: Optional[object_world_ids.FrameName] = None,
  ) -> data_types.Pose3:
    """Calls the ComputeIk rpc, doing argument conversion as necessary.

    If reference_frame is not specified, the WorldObject 'reference' will be
    used directly. The same applies for target_frame/target.

    Args:
      robot_name: Name of robot, must map to a kinematic object
      joints: the joint configuration values to compute fk for
      reference: object name for the reference of the returned pose
      target: object name for the target of the returned pose
      reference_frame: name that specifies a frame under 'reference', optional
      target_frame: name that specifies a frame under 'target', optional

    Returns:
      The reference_t_target pose, i.e. the pose of 'target' in the frame of
      'reference'
    """
    request = motion_planner_service_pb2.FkRequest(world_id=self._world_id)
    request.robot_reference.object_id.by_name.object_name = robot_name
    request.joints.joints.extend(joints)
    if reference_frame is not None:
      # If frame is not None, then we'll assume the user is specifying a frame,
      # and otherwise the user is specifying an object.
      request.reference.by_name.frame.object_name = reference
      request.reference.by_name.frame.frame_name = reference_frame
    else:
      request.reference.by_name.object.object_name = reference
    if target_frame is not None:
      # If frame is not None, then we'll assume the user is specifying a frame,
      # and otherwise the user is specifying an object.
      request.target.by_name.frame.object_name = target
      request.target.by_name.frame.frame_name = target_frame
    else:
      request.target.by_name.object.object_name = target

    # Make the rpc.
    response = self._stub.ComputeFk(request)
    return math_proto_conversion.pose_from_proto(response.reference_t_target)

  def check_collisions(
      self,
      robot_name: object_world_ids.WorldObjectName,
      waypoints: list[joint_space_pb2.JointVec],
      options: CheckCollisionsOptions,
  ) -> motion_planner_service_pb2.CheckCollisionsResponse:
    """Calls the CheckCollisions rpc.

    Args:
      robot_name: Name of robot, must map to a kinematic object.
      waypoints: The waypoints define the path for which we check the collision.
        We also check the linear interpolation between the waypoints.
      options: Check collision options

    Returns:
      CheckCollisionResponse. See flag `has_collision` in response
      which indicates a collision.
    """

    request = motion_planner_service_pb2.CheckCollisionsRequest(
        world_id=self._world_id,
        robot_reference=robot_specification_pb2.RobotReference(),
        waypoint=waypoints,
        collision_settings=options.collision_settings,
    )
    request.robot_reference.object_id.by_name.object_name = robot_name
    response = self._stub.CheckCollisions(request)
    return response

  def clear_cache(self) -> empty_pb2.Empty:
    """Calls the ClearCache rpc."""
    return self._stub.ClearCache(empty_pb2.Empty())
